internal/lang/eval.go

```diff
@@ -343,196 +340,253 @@ func (s *Scope) evalContext(parent *hcl.EvalContext, refs []*addrs.Reference, se
+	if !ok {
+		return diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  `Invalid "self" reference`,
+			// This detail message mentions some current practice that
+			// this codepath doesn't really "know about". If the "self"
+			// object starts being supported in more contexts later then
+			// we'll need to adjust this message.
+			Detail:  `The "self" object is not available in this context. This object can be used only in resource provisioner, connection, and postcondition blocks.`,
+			Subject: ref.SourceRange.ToHCL().Ptr(),
+		})
+	}
 
-		case addrs.LocalValue:
-			val, valDiags := normalizeRefValue(s.Data.GetLocalValue(subj, rng))
-			diags = diags.Append(valDiags)
-			localValues[subj.Name] = val
+	val, valDiags := normalizeRefValue(b.s.Data.GetResource(subj.ContainingResource(), ref.SourceRange))
+
+	diags = diags.Append(valDiags)
+
+	// Self is an exception in that it must always resolve to a
+	// particular instance. We will still insert the full resource into
+	// the context below.
+	var hclDiags hcl.Diagnostics
+	// We should always have a valid self index by this point, but in
+	// the case of an error, self may end up as a cty.DynamicValue.
+	switch k := subj.Key.(type) {
+	case addrs.IntKey:
+		b.self, hclDiags = hcl.Index(val, cty.NumberIntVal(int64(k)), ref.SourceRange.ToHCL().Ptr())
+	case addrs.StringKey:
+		b.self, hclDiags = hcl.Index(val, cty.StringVal(string(k)), ref.SourceRange.ToHCL().Ptr())
+	default:
+		b.self = val
+	}
+	diags = diags.Append(hclDiags)
 
-		case addrs.PathAttr:
-			val, valDiags := normalizeRefValue(s.Data.GetPathAttr(subj, rng))
-			diags = diags.Append(valDiags)
-			pathAttrs[subj.Name] = val
+	return diags
+}
 
-		case addrs.TerraformAttr:
-			val, valDiags := normalizeRefValue(s.Data.GetTerraformAttr(subj, rng))
-			diags = diags.Append(valDiags)
-			terraformAttrs[subj.Name] = val
+func (b *evalVarBuilder) putValueBySubject(ref *addrs.Reference) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
 
-		case addrs.CountAttr:
-			val, valDiags := normalizeRefValue(s.Data.GetCountAttr(subj, rng))
-			diags = diags.Append(valDiags)
-			countAttrs[subj.Name] = val
+	rawSubj := ref.Subject
+	rng := ref.SourceRange
+
+	// This type switch must cover all of the "Referenceable" implementations
+	// in package addrs, however we are removing the possibility of
+	// Instances beforehand.
+	switch addr := rawSubj.(type) {
+	case addrs.ResourceInstance:
+		rawSubj = addr.ContainingResource()
+	case addrs.ModuleCallInstance:
+		rawSubj = addr.Call
+	case addrs.ModuleCallInstanceOutput:
+		rawSubj = addr.Call.Call
+	}
 
-		case addrs.ForEachAttr:
-			val, valDiags := normalizeRefValue(s.Data.GetForEachAttr(subj, rng))
-			diags = diags.Append(valDiags)
-			forEachAttrs[subj.Name] = val
+	normalizeRefValue := func(val cty.Value, inDiags tfdiags.Diagnostics) cty.Value {
+		normV, normDiags := normalizeRefValue(val, inDiags)
+		diags = diags.Append(normDiags)
+		return normV
+	}
```

Comment:
Why not instead have a variable `normDiags` as a var in the `putValueBySubject` function, and then there would be no need for this sub-function, that mutates `diags`?

Reply:
Yep, that is much cleaner :)

[a5a9c69](https://github.com/opentofu/opentofu/pull/1750/commits/a5a9c696d3dbc411412a0a66deffc84f4a394d82)